### ISSUE TITLE
Specification for join fetching

### ISSUE THREAD

**Issue by geraldhumphries** [tkaczmarzyk/specification-arg-resolver#10]
Hi, I have a feature request for a spec class that will run a join fetch (ex. like `@Query("select distinct entity from Entity entity left join fetch entity.contacts")`) on a lazy loaded relationship. Using Specifications, there is no other way to emulate `join fetch` in `@Query`. I would rather the field be defined in the annotation as a constant, and it should not be required (or maybe not allowed) to pass it into the API.

I have created a working mockup here: https://gist.github.com/geraldhumphries/9fba746a4613aae78ca4

That is a hack, however. I think an annotation like `@JoinFetch(field = "lazyRelationship", joinType = JoinType.LEFT)` would work better.

**Comment by geraldhumphries** [comment]
I've updated the gist to support passing in multiple parameters. This can be done through request parameters or constVal. Works pretty well, what do you think?

**Comment by tkaczmarzyk** [comment]
The idea and the spec impl looks great! As far as the annotation, I think the best way is to have a custom one (like `@JoinFetch` that you proposed). I'm on it!

**Comment by tkaczmarzyk** [comment]
I added `@JoinFetch` which works with parameter specs as well as custom interfaces. It's pushed to `joinfetch` branch. Would you like to give it a try?

**Comment by geraldhumphries** [comment]
I have not had a chance to test it, but can the fetches be passed in through the API or are they constants-only?

I should have some time tomorrow, but it looks good so far!

**Comment by tkaczmarzyk** [comment]
It's constants only, I think it should be ok in most of the situations...

**Comment by geraldhumphries** [comment]
I just tested it, works great.

One suggestion, make it repeatable with the `@Repeatable` annotation so different fetches can be applied:

```
@And({@Spec(path = "status", spec = In.class),
    @Spec(path = "direction", spec = In.class),
}) @JoinFetch(paths = {"contacts", "clients"}, joinType = JoinType.LEFT)
@JoinFetch(paths = "segments", joinType = JoinType.INNER)
Specification<Load> loadSpec, ...
```

**Comment by tkaczmarzyk** [comment]
Hi again. Sorry for the delay. I added `@Joins` container annotation for repeating. I didn't make it `@Repeatable`, because it would require Java 8 which I want to avoid at this point.

It's now released as v0.7.0. Thank you very much for your help.

**Comment by geraldhumphries** [comment]
Awesome! Thanks.
